### ISSUE TITLE
Incomplete parsing of apt sources.list

### ISSUE THREAD

**Issue by candlerb** [ansible/ansible#1888]
The following is a valid entry in `/etc/apt/sources.list`; it is necessary because the repository in question doesn't have i386 packages and the packages are not signed.

```
deb [arch=amd64 trusted=yes] http://research.cs.wisc.edu/htcondor/debian/stable/ squeeze contrib
```

The sources.list(5) manpage describes this format.

However, ansible 0.9's apt module barfs on this entry:

```
TASK: [apt pkg=glusterfs-client state=latest update_cache=yes] ********************* 
fatal: [xxx.example.com] => failed to parse: Traceback (most recent call last):

  File "/root/.ansible/tmp/ansible-1358871677.86-33393517133507/apt", line 890, in <module>

    main()

  File "/root/.ansible/tmp/ansible-1358871677.86-33393517133507/apt", line 212, in main

    cache = apt.Cache()

  File "/usr/lib/python2.7/dist-packages/apt/cache.py", line 102, in __init__

    self.open(progress)

  File "/usr/lib/python2.7/dist-packages/apt/cache.py", line 149, in open

    self._list.read_main_list()

SystemError: E:Malformed line 67 in source list /etc/apt/sources.list (URI parse)
```

**Comment by dhozac** [comment]
That is coming from python-apt, which the apt module uses to interface with apt itself. You will want to report this to them, though I can't really figure out where to do that. A Debian bug might be best.

**Comment by candlerb** [comment]
FWIW this turned out to be a problem in apt-add-repository corrupting the sources.list file. Reported upstream as
https://bugs.launchpad.net/ubuntu/+source/software-properties/+bug/1103200

The way ansible's apt-repository module works, it calls apt-add-repository --remove followed by apt-add-repository every time it runs, even if you already have the repo installed.

**Comment by mpdehaan** [comment]
> The way ansible's apt-repository module works, it calls apt-add-repository
> --remove followed by apt-add-repository every time it runs, even if you
> already have the repo installed.

This really ought to be fixed.

**Comment by sfromm** [comment]
Does anyone have any objections to adding `apt` as a requirement to the apt_repository module?  Using that, i can then parse the sources.list and check to see if the repository is already configured.  That would remove the initial `apt-add-repository --remove` step.

**Comment by sfromm** [comment]
@candlerb, please try the patch in #1907 and let me know if that works for you.

**Comment by candlerb** [comment]
Yes it works for me (aside: to make the patch apply to 0.9 I had to change module.run_command to _run). Thanks!

**Comment by sfromm** [comment]
Merged e0741e7 to address the apt_repository module behavior described above.
